golf2/src/nvcounter_syscall.rs

```diff
@@ -0,0 +1,172 @@
+    /// will commence when it completes. If the initialization fails, the
+    /// counter will be poisoned and will become unable to operate. Worse, the
+    /// value stored in flash becomes undefined, although it will likely be a
+    /// value between 0 and the previous value.
+    #[allow(unused)]
+    pub fn initialize(&self) {
+        if self.nvcounter.initialize() != ReturnCode::SUCCESS {
+            self.handle_failed_init();
+        }
+    }
+
+    /// Sends failures to all apps with outstanding increment requests and marks
+    /// init_failed as true.
+    fn handle_failed_init(&self) {
+        self.init_failed.set(true);
+        self.grant.each(|app_data| {
+            if !app_data.wants_increment { return; }
+            app_data.wants_increment = false;
+            if let Some(mut callback) = app_data.callback {
+                callback.schedule(0, 0, 0);
+            }
+        });
+    }
+
+    // Scans through the apps and starts the next increment, if any app wants an
+    // increment. This will also call the callback for app callback_id with the
+    // given callback code -- specify an id if usize::max_value() if no callback
+    // is necessary.
+    fn do_next_op(&self, callback_id: usize, callback_code: usize) {
+        use ReturnCode::SuccessWithValue;
+        // TODO: Fairness? This seems to be the common approach but it gives
+        // priority to lower-numbered apps. Probably not an issue for this
+        // particular driver because read_and_increment() shouldn't see much
+        // contention.
+        self.grant.each(|app_data| {
+            if self.current_app.get() == usize::max_value() &&
+               app_data.wants_increment
+            {
+                app_data.wants_increment = false;
+                if let SuccessWithValue { value } =
+                    self.nvcounter.read_and_increment()
+                {
+                    self.value.set(value);
+                    self.current_app.set(app_data.appid().idx());
+                } else if let Some(mut callback) = app_data.callback {
+                    callback.schedule(0, 0, 0);
+                }
+            }
+
+            if app_data.appid().idx() == callback_id {
+                if let Some(mut callback) = app_data.callback {
+                    callback.schedule(callback_code, self.value.get(), 0);
+                }
+            }
+        });
+    }
+
+    fn read_and_increment(&self, app: AppId) -> ReturnCode {
+        if self.init_failed.get() { return ReturnCode::FAIL; }
+        let result = self.grant.enter(app, |app_data, _| {
+            if app_data.wants_increment { return ReturnCode::EBUSY; }
+            app_data.wants_increment = true;
+            ReturnCode::SUCCESS
+        }).unwrap_or(ReturnCode::ENOMEM);
+        if result != ReturnCode::SUCCESS { return result; }
+        if self.current_app.get() == usize::max_value() {
+            if self.nvcounter.read_and_increment() != ReturnCode::SUCCESS {
```

Comment:
Do you want to allow multiple apps to operate on the counter concurrently (i.e.. queue them)?

I think the answer is no -- don't add the complexity given we expect a single app right now -- but it would be good to document this.

Reply:
I virtualized it under the assumption that all syscall APIs should be virtualized. If you'd rather I remove the virtualization I'd be happy to do so.